### PR TITLE
[tune] More robust resolution/detection of signature

### PR DESCRIPTION
**Pull request by richardliaw**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/utils/util.py [hunk 1]
```diff
@@ -400,6 +400,50 @@ def validate_save_restore(trainable_cls,
...
+        raise ValueError(
+            "Provided training function must have 2 args "
+            "in the signature, and the latter arg must "
+            "contain `checkpoint_dir`. For example: "
+            "`func(config, checkpoint_dir=None)`. Got {}".format(func_args))
+    return validated
+
+
+def detect_reporter(func):
+    """Use checkpointing if any arg has "reporter" and args = 2"""
```
Comment (amogkam) [inline python/ray/tune/utils/util.py:424]:
```suggestion
    """Use reporter if any arg has "reporter" and args = 2"""
```

File: python/ray/tune/utils/util.py [hunk 2]
```diff
@@ -400,6 +400,50 @@ def validate_save_restore(trainable_cls,
...
+    try:
+        func_sig.bind({}, reporter=None)
+    except Exception as e:
+        logger.debug(str(e))
+        use_reporter = False
+    return use_reporter
+
+
+def detect_config_single(func):
+    """Check if func({}) workers."""
```
Comment (amogkam) [inline python/ray/tune/utils/util.py:436]: Is this a typo?
```suggestion
    """Check if func({}) works."""
```

File: python/ray/tune/function_runner.py [hunk 3]
```diff
@@ -485,16 +469,16 @@ def wrap_function(train_func, warn=True):
...
-    if len(func_args) > 1:  # more arguments than just the config
-        if "reporter" not in func_args and not use_checkpoint:
-            raise ValueError(
-                "Unknown argument found in the Trainable function. "
-                "Arguments other than the 'config' arg must be one "
-                "of ['reporter', 'checkpoint_dir']. Found: {}".format(
-                    func_args))
+    use_config_single = detect_config_single(train_func)
+    use_reporter = detect_reporter(train_func)
+
```
Comment (amogkam) [inline python/ray/tune/function_runner.py:474]: Isn't `train_func(config, checkpoint_dir, reporter)` valid as well? Should there be a third case detecting when both checkpoint_dir and reporter are passed in?

**Review by amogkam — COMMENTED** [review]
Nice! Just left a couple comments.

File: python/ray/tune/function_runner.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/tune/function_runner.py:474]: No, `reporter` shouldn't work with checkpoint_dir (intentionally). Actually, we should just remove the option to use a reporter.

```
In [1]: def tune_(config, reporter, checkpoint_dir):
   ...:     pass
   ...:

In [2]: from ray import tune

In [3]: tune.run(tune_)

ray.tune.error.TuneError: Trial raised an exception. Traceback:
ray::ImplicitFunc.train() (pid=5960, ip=192.168.1.115)
  File "/Users/rliaw/miniconda3/lib/python3.7/site-packages/ray/tune/function_runner.py", line 225, in run
    self._entrypoint()
  File "/Users/rliaw/miniconda3/lib/python3.7/site-packages/ray/tune/function_runner.py", line 288, in entrypoint
    self._status_reporter.get_checkpoint())
  File "/Users/rliaw/miniconda3/lib/python3.7/site-packages/ray/tune/function_runner.py", line 515, in _trainable_func
    output = train_func(config, reporter)
```

**Review by krfricke — APPROVED** [review]
LGTM

**Review by amogkam — APPROVED** [review]

**Pull request merged**